pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java

```diff
@@ -222,6 +233,42 @@ public void handleEndTxnOnSubscriptionResponse(long requestId,
         }
     }
 
+    private void continueCheckOutstandingRequestIfTimeout() {
+        Map.Entry<Long, OpRequestSend> entry = outstandingRequests.firstEntry();
+        if (entry == null) {
+            timoutCheckIsRunning.set(false);
+            // To avoid race conditions, the scenarios are as follows.
+            // Thread A: after if(entry=null) => Thread B: put a new entry and then get timoutCheckIsRunning=true =>
+            // Thread A: set timoutCheckIsRunning = false
+            if (outstandingRequests.firstEntry() != null
+                    && timoutCheckIsRunning.compareAndSet(false, true)) {
+                 continueCheckOutstandingRequestIfTimeout();
+            }
+            return;
+        }
+        OpRequestSend op = entry.getValue();
+        final long finalRequestId = op.requestId;
+        if (!op.isTimeout(operationTimeoutInMills)) {
```

Comment:
If this op have been completed by handle response, the timeout time is not correct